poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+ *    secret shares are valid.
+ *
+ * 4) If the secret share is invalid, an insurer creates a BlameProof and sends
+ *    it back.
+ *
+ * 5) If the share is valid, an insurer creates a PromiseSignature to send back
+ *    to the promiser.
+ *
+ * 6) The promiser receives the message from the insurer.
+ *
+ * 7) If it is a valid BlameProof, the promiser must start all over and
+ *    construct a non-malicious Promise (or, the system can ban this malicious
+ *    promiser).
+ *
+ * 8) If the message is a PromiseSignature, the promiser can add the signature
+ *    to its PromiseState
+ *
+ * 9) Steps 2-8 are repeated until the promiser has collected enough
+ *    PromiseSignatures for the Promise to be considered certified.
+ *
+ * 10) Once the Promise is certified, the promiser can then send the promise to
+ *     clients.
+ *
+ * 11) Clients can then request the signatures from the insurers to make sure
+ *     the Promise is indeed certified
+ *
+ *      i) This prevents a malicious promiser from simply leaving out valid
+ *         BlameProofs and only sending good signatures to the clients.
+ *
+ * 12) The promiser can perform some work for clients.
+ *
+ * 13) If the promiser is unresponsive for too long, a client can inform the
+ *     insurers of the Promise.
+ *
+ * 14) The insurers can then check if the promiser is indeed unresponsive.
+ *
+ * 15) If so, an insurer reveal its share and send it to the client.
+ *
+ * 16) The client repeats steps 13-15 until enough shares are recovered to
+ *     reconstruct the secret.
+ *
+ * 17) The client can then reconstruct the secret and take over for the promiser
+ *
+ *
+ *
+ * Key Terms:
+ *   promiser = the server making a Promise
+ *   client   = recipients of a Promise who are trusting the promiser
+ *   insurer  = servers who store secret shares of a promise. Such servers help
+ *              during secret reconstruction.
+ *
+ *   Users of this code = programmers wishing to use this code in programs
+*/
+
+// Used mostly in marshalling code, this is the size of a uint32
+var uint32Size int = binary.Size(uint32(0))
+
+// This is the protocol name used by crypto/proof verifiers and provers.
+var protocolName string = "Promise Protocol"
+
+// These are messages used for signatures
+var sigMsg []byte = []byte("Promise Signature")
+var sigBlameMsg []byte = []byte("Promise Blame Signature")
+
+/* Promise structs are mechanisms by which a server can promise other servers
+ * that an abstract.Secret will be availble even if the secret's owner goes
+ * down. The secret to be promised will be sharded into shared secrets that can
+ * be combined using Lagrange Interpolation to produce the original secret.
+ * Other servers will act as insurers maintaining a share. If a client ever
+ * needs the secret to be reconstructed, it can contact the insurers to regain
+ * the shares and reconstruct the secret.
+ *
+ * Note to Developers:
```

Comment:
The text "note to developers" is redundant since it is only developers who tend to read API documentation like this.